src/librustc_trans/abi.rs

```diff
@@ -127,7 +127,12 @@ impl LlvmType for Reg {
 impl LlvmType for CastTarget {
     fn llvm_type(&self, cx: &CodegenCx) -> Type {
         let rest_ll_unit = self.rest.unit.llvm_type(cx);
-        let rest_count = self.rest.total.bytes() / self.rest.unit.size.bytes();
+        let rest_count = if self.rest.unit.size.bytes() == 0 {
+            0
+        } else {
+            self.rest.total.bytes() / self.rest.unit.size.bytes()
+        };
+
         let rem_bytes = self.rest.total.bytes() % self.rest.unit.size.bytes();
```

Comment:
If `y == 0`, then `x % y` will also panic, please move this `rem_bytes` calculation inside the conditional as well.